http/tus.go

```diff
@@ -0,0 +1,271 @@
+package http
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"sync"
+
+	"github.com/tus/tusd/pkg/filestore"
+	tusd "github.com/tus/tusd/pkg/handler"
+
+	"github.com/filebrowser/filebrowser/v2/settings"
+	"github.com/filebrowser/filebrowser/v2/storage"
+	"github.com/filebrowser/filebrowser/v2/users"
+)
+
+const uploadDirName = ".tmp_upload"
+
+type TusHandler struct {
+	store                *storage.Storage
+	server               *settings.Server
+	settings             *settings.Settings
+	tusdHandlers         map[uint]*tusd.UnroutedHandler
+	notifyNewTusdHandler chan struct{}
+	apiPath              string
+	mutex                *sync.Mutex
+}
+
+func NewTusHandler(store *storage.Storage, server *settings.Server, apiPath string) (TusHandler, error) {
+	tusHandler := TusHandler{}
+	tusHandler.store = store
+	tusHandler.server = server
+	tusHandler.tusdHandlers = make(map[uint]*tusd.UnroutedHandler)
+	tusHandler.notifyNewTusdHandler = make(chan struct{})
+	tusHandler.apiPath = apiPath
+	tusHandler.mutex = &sync.Mutex{}
+
+	var err error
+	if tusHandler.settings, err = store.Settings.Get(); err != nil {
+		return tusHandler, fmt.Errorf("couldn't get settings: %w", err)
+	}
+
+	// Create a goroutine that handles uploaded file events for all users
+	go tusHandler.handleFileUploadedEvents()
+
+	return tusHandler, nil
+}
+
+func (th TusHandler) getOrCreateTusdHandler(d *data, r *http.Request) (*tusd.UnroutedHandler, error) {
+	// Use a mutex to make sure only one tus handler is created for each user
+	th.mutex.Lock()
+	defer th.mutex.Unlock()
+
+	tusdHandler, ok := th.tusdHandlers[d.user.ID]
+	log.Printf("Getting tus handler for user %s with basePath %s\n", d.user.Username, d.user.FullPath("/"))
+	if !ok {
+		// If we don't define an absolute URL for tusd, it creates an absolute URL for us that the client will use.
+		// See tusd/handler/unrouted_handler.go/absFileURL() for details.
+		// This URL's scheme will be http in our case (as we don't use tusd's inbuilt TLS feature),
+		// which is fine if we don't use both a browser and a reverse proxy that terminates SSL for us.
+		// In case we do, we need to define an absolute URL with the correct scheme, or we'll get mixed content errors.
+		// We can extract the correct scheme and host from the origin request header, if it exists (which always is the case for browsers).
+		var origin string
+		if originHeader, ok := r.Header["Origin"]; ok && len(originHeader) > 0 {
+			origin = originHeader[0]
+		}
+		basePath, err := url.JoinPath(origin, th.server.BaseURL, th.apiPath)
+		if err != nil {
+			return nil, err
+		}
+
+		log.Printf("Creating tus handler for user %s on path %s\n", d.user.Username, basePath)
+		tusdHandler, err := th.createTusdHandler(d, basePath) //nolint:govet
```

Comment:
This code is a bit confusing and might be error-prone because it `tusdHandler` shadows the variable created in the surrounding scope. As a workaround, var err error can be defined or a function definition can be changed to  

`getOrCreateTusdHandler(d *data, r *http.Request) (_ *tusd.UnroutedHandler, err error)`

```suggestion
tusdHandler, err = th.createTusdHandler(d, basePath)
```